### PR TITLE
COCO SeqFile

### PR DESCRIPTION
**Pull request by Menooker**
## What changes were proposed in this pull request?

Split the PR #2917 into this PR and  #2913 . This PR only contains COCO SeqFile loading and writing.

## How was this patch tested?

UT

## Related links or issues (optional)
https://github.com/intel-analytics/BigDL/pull/2917

### CODE REVIEW

**Comment by zhangxiaoli73** [comment]
add unit tests

**Comment by zhangxiaoli73** [comment]
please rebase your code

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/DataSet.scala [hunk 1]
```diff
@@ -586,6 +589,48 @@ object DataSet {
...
+          val inputStream = new ByteArrayInputStream(data._2.getBytes)
+          val image = ImageIO.read(inputStream)
+          val rawdata = image.getRaster.getDataBuffer.asInstanceOf[DataBufferByte].getData()
+          val imf = ImageFeature(rawdata, RoiLabel(labelClasses, bboxes, masks), fileName)
+          imf(ImageFeature.originalSize) = (height, width, 3)
+          imf(RoiLabel.ISCROWD) = isCrowd
+          imf
+        }
+        .coalesce(num)
+      ImageFrame.rdd(rawData)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/DataSet.scala]: Need bug fix

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/COCODataset.scala [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+   extends RLEMasks(_counts, _height, _width) with COCOSegmentation {
+   override def dumpTo(context: COCOSerializeContext): Unit = {
+     context.dump(counts.length)
+     counts.foreach(p => {
+       context.dump(p)
+     })
+   }
+ }
+
+object COCODataset {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/COCODataset.scala]: So how to get whole COCO dataset from this object?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/COCOSeqFileGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    parser.parse(args, COCOSeqFileGeneratorParams()).foreach { param =>
+      println("Loading COCO metadata")
+      val meta = COCODataset.load(param.metaPath, param.folder)
+      println("Metadata loaded")
+      val conf: Configuration = new Configuration
+      val doneCount = new AtomicInteger(0)
+      val tasks = meta.images.filter(img => {
+        val path = img.path
+        val valid = Files.exists(path) && !Files.isDirectory(path)
+        if (!valid) {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/COCOSeqFileGenerator.scala]: ok, good

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala [hunk 4]
```diff
@@ -31,15 +31,17 @@ abstract class SegmentationMasks extends Serializable {
...
  * @param poly An array of polygons. The inner array defines one polygon, with [x1,y1,x2,y2,...]
- * @param height the height of the image
- * @param width the width of the image
+ * @param _height the height of the image
+ * @param _width the width of the image
  */
-class PolyMasks(val poly: Array[Array[Float]], val height: Int, val width: Int) extends
-  SegmentationMasks {
+class PolyMasks(val poly: Array[Array[Float]],
+  private val _height: Int, private val _width: Int) extends SegmentationMasks {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala]: not change

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/DataSet.scala (hunk 1 above)
Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/DataSet.scala]: Changed to Dataset[ImageFeature]

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/COCODataset.scala (hunk 2 above)
Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/COCODataset.scala]: Can now load data by COCOImage

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala (hunk 4 above)
Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/segmentation/MaskUtils.scala]: changed back

**Comment by Menooker** [comment]
Jenkins pass http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/974/

**Comment by wzhongyuan** [comment]
@Menooker @zhangxiaoli73   we should have an example with doc how to do validation

**Comment by Menooker** [comment]
maybe in another PR? This pr has only coco dataset related code...

**Pull request merged**